File: ios/Live Target/CaliberData.swift
//
//  CaliberData.swift
//  Live Target
//
//  Created by Bryan Everly on 7/11/25.
//  Copyright © 2025 BCEAssociates, Inc. All rights reserved.
//

import Foundation

struct Caliber: Identifiable, Hashable {
    let id = UUID()
    let name: String
    let category: String
    let diameterInches: Double
    let common: Bool
    
    var displayName: String {
        return name
    }
    
    var pixelSize: Int {
        return Int(diameterInches * 100)
    }
}

class CaliberData: ObservableObject {
    static let shared = CaliberData()
    
    let calibers: [Caliber] = [
        Caliber(name: ".17 HMR", category: "Rimfire", diameterInches: 0.172, common: true),
        Caliber(name: ".17 Remington", category: "Centerfire Rifle", diameterInches: 0.172, common: false),
        Caliber(name: ".22 Short", category: "Rimfire", diameterInches: 0.223, common: true),
        Caliber(name: ".22 Long Rifle", category: "Rimfire", diameterInches: 0.223, common: true),
        Caliber(name: ".22 WMR", category: "Rimfire", diameterInches: 0.224, common: true),
        Caliber(name: ".22 Hornet", category: "Centerfire Rifle", diameterInches: 0.224, common: false),
        Caliber(name: ".223 Remington", category: "Centerfire Rifle", diameterInches: 0.224, common: true),
        Caliber(name: "5.56×45mm NATO", category: "Centerfire Rifle", diameterInches: 0.224, common: true),
        Caliber(name: ".243 Winchester", category: "Centerfire Rifle", diameterInches: 0.243, common: true),
        Caliber(name: "6mm Remington", category: "Centerfire Rifle", diameterInches: 0.243, common: false),
        Caliber(name: "6.5 Carcano", category: "Centerfire Rifle", diameterInches: 0.268, common: false),
        Caliber(name: "6.5×55mm Swedish", category: "Centerfire Rifle", diameterInches: 0.264, common: false),
        Caliber(name: "6.5 Creedmoor", category: "Centerfire Rifle", diameterInches: 0.264, common: true),
        Caliber(name: "6.5 Grendel", category: "Centerfire Rifle", diameterInches: 0.264, common: false),
        Caliber(name: ".270 Winchester", category: "Centerfire Rifle", diameterInches: 0.277, common: true),
        Caliber(name: "7mm Remington Magnum", category: "Centerfire Rifle", diameterInches: 0.284, common: true),
        Caliber(name: "7mm-08 Remington", category: "Centerfire Rifle", diameterInches: 0.284, common: false),
        Caliber(name: "7×57mm Mauser", category: "Centerfire Rifle", diameterInches: 0.284, common: false),
        Caliber(name: ".30 Carbine", category: "Centerfire Rifle", diameterInches: 0.308, common: false),
        Caliber(name: ".30-30 Winchester", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: ".308 Winchester", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: "7.62×51mm NATO", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: ".30-06 Springfield", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: "7.62×39mm", category: "Centerfire Rifle", diameterInches: 0.311, common: true),
        Caliber(name: "7.62×54mmR", category: "Centerfire Rifle", diameterInches: 0.311, common: false),
        Caliber(name: ".300 Winchester Magnum", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: ".300 Blackout", category: "Centerfire Rifle", diameterInches: 0.308, common: true),
        Caliber(name: "8mm Mauser", category: "Centerfire Rifle", diameterInches: 0.323, common: false),
        Caliber(name: ".338 Lapua Magnum", category: "Centerfire Rifle", diameterInches: 0.338, common: false),
        Caliber(name: ".25 ACP", category: "Centerfire Pistol", diameterInches: 0.251, common: false),
        Caliber(name: ".32 ACP", category: "Centerfire Pistol", diameterInches: 0.312, common: false),
        Caliber(name: ".380 ACP", category: "Centerfire Pistol", diameterInches: 0.355, common: true),
        Caliber(name: "9mm Luger", category: "Centerfire Pistol", diameterInches: 0.355, common: true),
        Caliber(name: "9×18mm Makarov", category: "Centerfire Pistol", diameterInches: 0.365, common: false),
        Caliber(name: ".38 Special", category: "Centerfire Pistol", diameterInches: 0.357, common: true),
        Caliber(name: ".357 Magnum", category: "Centerfire Pistol", diameterInches: 0.357, common: true),
        Caliber(name: ".40 S&W", category: "Centerfire Pistol", diameterInches: 0.400, common: true),
        Caliber(name: "10mm Auto", category: "Centerfire Pistol", diameterInches: 0.400, common: false),
        Caliber(name: ".44 Special", category: "Centerfire Pistol", diameterInches: 0.429, common: false),
        Caliber(name: ".44 Magnum", category: "Centerfire Pistol", diameterInches: 0.429, common: true),
        Caliber(name: ".45 ACP", category: "Centerfire Pistol", diameterInches: 0.452, common: true),
        Caliber(name: ".45 Colt", category: "Centerfire Pistol", diameterInches: 0.452, common: false),
        Caliber(name: ".50 AE", category: "Centerfire Pistol", diameterInches: 0.500, common: false),
        Caliber(name: ".410 Bore", category: "Shotgun", diameterInches: 0.410, common: false),
        Caliber(name: "20 Gauge", category: "Shotgun", diameterInches: 0.615, common: false),
        Caliber(name: "16 Gauge", category: "Shotgun", diameterInches: 0.670, common: false),
        Caliber(name: "12 Gauge", category: "Shotgun", diameterInches: 0.729, common: false),
        Caliber(name: "10 Gauge", category: "Shotgun", diameterInches: 0.775, common: false)
    ]
    
    var commonCalibers: [Caliber] {
        return calibers.filter { $0.common }.sorted { $0.name < $1.name }
    }
    
    var allCalibers: [Caliber] {
        return calibers.sorted { $0.name < $1.name }
    }
    
    var pistolCalibers: [Caliber] {
        return calibers.filter { $0.category == "Centerfire Pistol" }.sorted { $0.name < $1.name }
    }
    
    var rifleCalibers: [Caliber] {
        return calibers.filter { $0.category.contains("Rifle") }.sorted { $0.name < $1.name }
    }
    
    var rimfireCalibers: [Caliber] {
        return calibers.filter { $0.category == "Rimfire" }.sorted { $0.name < $1.name }
    }
    
    func findCaliber(byName name: String) -> Caliber? {
        return calibers.first { $0.name == name }
    }
    
    func findCaliber(byDiameter diameter: Double) -> Caliber? {
        return calibers.first { abs($0.diameterInches - diameter) < 0.001 }
    }
}

File: android/app/src/main/java/com/bceassociates/livetarget/data/CaliberData.kt
package com.bceassociates.livetarget.data

data class Caliber(
    val name: String,
    val category: String,
    val diameterInches: Double,
    val common: Boolean,
) {
    val pixelSize: Int
        get() = (diameterInches * 100).toInt()
}

object CaliberData {
    val calibers =
        listOf(
            Caliber(".17 HMR", "Rimfire", 0.172, true),
            Caliber(".17 Remington", "Centerfire Rifle", 0.172, false),
            Caliber(".22 Short", "Rimfire", 0.223, true),
            Caliber(".22 Long Rifle", "Rimfire", 0.223, true),
            Caliber(".22 WMR", "Rimfire", 0.224, true),
            Caliber(".22 Hornet", "Centerfire Rifle", 0.224, false),
            Caliber(".223 Remington", "Centerfire Rifle", 0.224, true),
            Caliber("5.56×45mm NATO", "Centerfire Rifle", 0.224, true),
            Caliber(".243 Winchester", "Centerfire Rifle", 0.243, true),
            Caliber("6mm Remington", "Centerfire Rifle", 0.243, false),
            Caliber("6.5 Carcano", "Centerfire Rifle", 0.268, false),
            Caliber("6.5×55mm Swedish", "Centerfire Rifle", 0.264, false),
            Caliber("6.5 Creedmoor", "Centerfire Rifle", 0.264, true),
            Caliber("6.5 Grendel", "Centerfire Rifle", 0.264, false),
            Caliber(".270 Winchester", "Centerfire Rifle", 0.277, true),
            Caliber("7mm Remington Magnum", "Centerfire Rifle", 0.284, true),
            Caliber("7mm-08 Remington", "Centerfire Rifle", 0.284, false),
            Caliber("7×57mm Mauser", "Centerfire Rifle", 0.284, false),
            Caliber(".30 Carbine", "Centerfire Rifle", 0.308, false),
            Caliber(".30-30 Winchester", "Centerfire Rifle", 0.308, true),
            Caliber(".308 Winchester", "Centerfire Rifle", 0.308, true),
            Caliber("7.62×51mm NATO", "Centerfire Rifle", 0.308, true),
            Caliber(".30-06 Springfield", "Centerfire Rifle", 0.308, true),
            Caliber("7.62×39mm", "Centerfire Rifle", 0.311, true),
            Caliber("7.62×54mmR", "Centerfire Rifle", 0.311, false),
            Caliber(".300 Winchester Magnum", "Centerfire Rifle", 0.308, true),
            Caliber(".300 Blackout", "Centerfire Rifle", 0.308, true),
            Caliber("8mm Mauser", "Centerfire Rifle", 0.323, false),
            Caliber(".338 Lapua Magnum", "Centerfire Rifle", 0.338, false),
            Caliber(".25 ACP", "Centerfire Pistol", 0.251, false),
            Caliber(".32 ACP", "Centerfire Pistol", 0.312, false),
            Caliber(".380 ACP", "Centerfire Pistol", 0.355, true),
            Caliber("9mm Luger", "Centerfire Pistol", 0.355, true),
            Caliber("9×18mm Makarov", "Centerfire Pistol", 0.365, false),
            Caliber(".38 Special", "Centerfire Pistol", 0.357, true),
            Caliber(".357 Magnum", "Centerfire Pistol", 0.357, true),
            Caliber(".40 S&W", "Centerfire Pistol", 0.400, true),
            Caliber("10mm Auto", "Centerfire Pistol", 0.400, false),
            Caliber(".44 Special", "Centerfire Pistol", 0.429, false),
            Caliber(".44 Magnum", "Centerfire Pistol", 0.429, true),
            Caliber(".45 ACP", "Centerfire Pistol", 0.452, true),
            Caliber(".45 Colt", "Centerfire Pistol", 0.452, false),
            Caliber(".50 AE", "Centerfire Pistol", 0.500, false),
            Caliber(".410 Bore", "Shotgun", 0.410, false),
            Caliber("20 Gauge", "Shotgun", 0.615, false),
            Caliber("16 Gauge", "Shotgun", 0.670, false),
            Caliber("12 Gauge", "Shotgun", 0.729, false),
            Caliber("10 Gauge", "Shotgun", 0.775, false),
        )

    val commonCalibers: List<Caliber>
        get() = calibers.filter { it.common }.sortedBy { it.name }

    val allCalibers: List<Caliber>
        get() = calibers.sortedBy { it.name }

    val pistolCalibers: List<Caliber>
        get() = calibers.filter { it.category == "Centerfire Pistol" }.sortedBy { it.name }

    val rifleCalibers: List<Caliber>
        get() = calibers.filter { it.category.contains("Rifle") }.sortedBy { it.name }

    val rimfireCalibers: List<Caliber>
        get() = calibers.filter { it.category == "Rimfire" }.sortedBy { it.name }

    val shotgunCalibers: List<Caliber>
        get() = calibers.filter { it.category == "Shotgun" }.sortedBy { it.name }

    fun findCaliberByName(name: String): Caliber? {
        return calibers.find { it.name == name }
    }

    fun findCaliberByDiameter(diameter: Double): Caliber? {
        return calibers.find { kotlin.math.abs(it.diameterInches - diameter) < 0.001 }
    }
}
